### PR TITLE
DRILL-7985: Support Mongo aggregate, union, project, limit, sort pushdowns

### PR DESCRIPTION
**Pull request by vvysotskyi**
# [DRILL-7985](https://issues.apache.org/jira/browse/DRILL-7985): Support Mongo aggregate, union, project, limit, sort pushdowns

## Description
This pull request adds support for the following mongo pushdowns:
- aggregate
- project (including expressions)
- limit (including offset)
- union / union all
- sort
Added storage plugin configs to disable specific optimizations if required.

Besides that, it includes changes for the Generic storage plugin pushdown framework that simplifies the implementation of specific pushdowns for the planning stage. For more details about it, please refer to this doc: https://github.com/apache/drill/compare/master...vvysotskyi:DRILL-7985?expand=1#diff-b0e0fdc4f6e47614a62fbfdbad0a21598726c72d15de22120e4ac9dbd945e933
It works in a similar way to how we do pushdowns for JDBC, Cassandra, or Elasticsearch, but without using the Calcite adapter.

## Documentation
Yes

## Testing
Added UT.

### CODE REVIEW

**Comment by vvysotskyi** [comment]
Hi @julianhyde, part of this pull request introduces an API for adding support of pushdowns to the storage plugins, and it is something similar to what Calcite adapters do. Do you think it makes sense to move it to the Calcite, rewrite adapters to use this common approach, and make them more extensible, so we will be able to customize them in Drill and query using the Drill execution engine?

**Comment by julianhyde** [comment]
@vvysotskyi Of course it depends what is feasible for Drill, given its current APIs, users, and code base.

But I introduced Calcite's model of adapters and am still a big believer. (Indeed I tried to convince Drill to use Calcite's model in the early days.) I observed that if you build a simple adapter that supports scan, you have to add another parameter for filter, a way to handle rejected filters (or parts of filters), add another parameter to push down expressions and project columns (and handle rejected expressions), and it becomes exponentially more complex as you add additional operations (aggregate, join, sort, limit).

So I proposed the model where you write rewrite rules, each of which matches a particular relational operator. The rules tend to compose fairly well.

Maybe Drill could keep its existing Mongo adapter but also support generic Calcite adapters (powered by Calcite rewrite rules). Then Drill could pull in Calcite's Mongo adapter. If there are use cases that Calcite's Mongo adapter does not handle, they could be dealt with as Calcite bugs.

**Comment by cgivre** [comment]
> @vvysotskyi Of course it depends what is feasible for Drill, given its current APIs, users, and code base.
> 
> But I introduced Calcite's model of adapters and am still a big believer. (Indeed I tried to convince Drill to use Calcite's model in the early days.) I observed that if you build a simple adapter that supports scan, you have to add another parameter for filter, a way to handle rejected filters (or parts of filters), add another parameter to push down expressions and project columns (and handle rejected expressions), and it becomes exponentially more complex as you add additional operations (aggregate, join, sort, limit).
> 
> So I proposed the model where you write rewrite rules, each of which matches a particular relational operator. The rules tend to compose fairly well.
> 
> Maybe Drill could keep its existing Mongo adapter but also support generic Calcite adapters (powered by Calcite rewrite rules). Then Drill could pull in Calcite's Mongo adapter. If there are use cases that Calcite's Mongo adapter does not handle, they could be dealt with as Calcite bugs.

@julianhyde Thanks for the response!  @vvysotskyi  One other thing to consider is that Drill needs these rules for query planning and parallelization.  From my recollection, that was the major challenge in using the Calcite adapters as written.  IMHO it would be great if we could find a way to accomplish both goals using the Calcite adapters.

**Comment by julianhyde** [comment]
Note that Calcite's "adapter" abstraction isn't as strong as Drill's. It's just a bunch of rules (and often a calling convention). So maybe you can just pull in the rules you need.

**Comment by vvysotskyi** [comment]
Calcite rules for adapters in most of the cases have parts of the logic related to the storage itself. My idea was to move out this logic from rules to "implementor", so we will have simple rules that could be reused, and adapters will provide only the "implementor" that will be responsible for checking whether the specific pushdown can be done and handling it.

**Review by paul-rogers — COMMENTED** [review]
Nice addition. Although I was asked to review this, my knowledge of the Calcite side of things is limited, so I'll defer to others on this one.

File: contrib/storage-mongo/src/main/java/org/apache/drill/exec/store/mongo/MongoRecordReader.java [hunk 1]
```diff
@@ -176,13 +185,15 @@ public int next() {
...
+      MongoIterable<BsonDocument> projection;
+      if (CollectionUtils.isNotEmpty(operations)) {
+        List<Bson> operations = new ArrayList<>(this.operations);
+        operations.add(Aggregates.project(fields));
+        projection = collection.aggregate(operations);
       } else {
-        cursor = collection.find(filters).projection(fields).batchSize(100).iterator();
+        projection = collection.find(filters).projection(fields);
       }
+      cursor = projection.batchSize(100).iterator();
```
Comment (cgivre) [inline contrib/storage-mongo/src/main/java/org/apache/drill/exec/store/mongo/MongoRecordReader.java:196]: Is there a reason the `batchSize` is `100`?  Could we set this as a constant?  Or alternatively, a configurable option?

File: contrib/storage-mongo/src/main/java/org/apache/drill/exec/store/mongo/MongoRecordReader.java [hunk 2]
```diff
@@ -87,15 +92,19 @@ public MongoRecordReader(MongoSubScan.MongoSubScanSpec subScanSpec, List<SchemaP
...
+          shardedMongoSubScanSpec.getMinFilters(), shardedMongoSubScanSpec.getMaxFilters());
 
-    buildFilters(subScanSpec.getFilter(), mergedFilters);
+      buildFilters(shardedMongoSubScanSpec.getFilter(), mergedFilters);
+    }
     enableAllTextMode = fragmentContext.getOptions().getOption(ExecConstants.MONGO_ALL_TEXT_MODE).bool_val;
     enableNanInf = fragmentContext.getOptions().getOption(ExecConstants.JSON_READER_NAN_INF_NUMBERS).bool_val;
     readNumbersAsDouble = fragmentContext.getOptions().getOption(ExecConstants.MONGO_READER_READ_NUMBERS_AS_DOUBLE).bool_val;
     isBsonRecordReader = fragmentContext.getOptions().getOption(ExecConstants.MONGO_BSON_RECORD_READER).bool_val;
-    maxRecords = subScanSpec.getMaxRecords();
```
Comment (cgivre) [inline contrib/storage-mongo/src/main/java/org/apache/drill/exec/store/mongo/MongoRecordReader.java:98]: Are we applying the limit somewhere else?

File: contrib/storage-mongo/src/main/java/org/apache/drill/exec/store/mongo/plan/MongoPluginImplementor.java [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+    } else {
+      filters = mongoFilterBuilder.parseTree();
+    }
+  }
+
+  @Override
+  public void implement(PluginLimitRel limit) throws IOException {
+    runAggregate = true;
+    visitChild(limit.getInput());
+
```
Comment (cgivre) [inline contrib/storage-mongo/src/main/java/org/apache/drill/exec/store/mongo/plan/MongoPluginImplementor.java:123]: I'm a little confused here.  What happens if the operation isn't an aggregate?  Do we still apply the limit somewhere?

**Review by cgivre — COMMENTED** [review]
@vvysotskyi 
Thanks for this.  This is a really huge improvement to the pushdown architecture.  I had a few questions for you left in the comments.

**Review by vvysotskyi — COMMENTED** [review]
@cgivre, thanks for the review! I've addressed CR comments, please take a look.

**Review by cgivre — COMMENTED** [review]
@vvysotskyi 
This looks really good and thanks for the revisions.  I have one question about authentication.  I see that we added the credential provider to the Mongo config, but I don't see how a user would actually provide credentials to Drill.  (I'm not that familiar with Mongo, so forgive me if this is a really obvious question)

Are the creds provided in the URL?  If so, can a user put the connection URL in the credentialProvider?   Could we add a short `README` documenting this?

Everything else looks good.  I'm really liking the lombok improvements.

**Comment by vvysotskyi** [comment]
@cgivre, it is a good question, I've added a doc with this info in a new commit, please take a look.

**Review by cgivre — COMMENTED** [review]
@vvysotskyi Thanks for the docs.  The docs look good, but where are the creds being added to the connection URL?  In the ES plugin I see the method below which does that.  

https://github.com/apache/drill/blob/41c2dcc498afcbac8aa584c4180445ebf62b34ff/contrib/storage-elasticsearch/src/main/java/org/apache/drill/exec/store/elasticsearch/ElasticsearchStorageConfig.java#L78-L86

**Comment by cgivre** [comment]
One other thing... would it be worthwhile to update the Mongo Driver to the latest version (4.3.1), or should we save that for another PR?

**Comment by vvysotskyi** [comment]
@cgivre, here it is: 
https://github.com/apache/drill/blob/41c2dcc498afcbac8aa584c4180445ebf62b34ff/contrib/storage-mongo/src/main/java/org/apache/drill/exec/store/mongo/MongoStoragePlugin.java#L83-L101

Thanks for pointing to the newer driver version, I have updated it.

**Review by cgivre — APPROVED** [review]
LGTM +1
Thank you very much for this much needed improvement.  I think the refactoring will go a long way into making storage plugins more effective and more maintainable.

File: contrib/storage-mongo/src/main/java/org/apache/drill/exec/store/mongo/plan/RexToMongoTranslator.java [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+      if (op1 instanceof RexLiteral) {
+        if (op1.getType().getSqlTypeName() == SqlTypeName.INTEGER) {
+          return new BsonDocument("$arrayElemAt", new BsonArray(
+              Arrays.asList(strings.get(0), new BsonInt32(((RexLiteral) op1).getValueAs(Integer.class)))));
+        } else if (op1.getType().getSqlTypeName() == SqlTypeName.CHAR) {
+          return new BsonString(strings.get(0).asString().getValue() + "." + ((RexLiteral) op1).getValueAs(String.class));
+        }
+      }
+    }
+    if (call.getOperator() == SqlStdOperatorTable.CASE) {
```
Comment (jaltekruse) [inline contrib/storage-mongo/src/main/java/org/apache/drill/exec/store/mongo/plan/RexToMongoTranslator.java:157]: It doesn't look like there is a test for this multi-leg case translation, can we add one?

**Review by vdiravka — COMMENTED** [review]
We are adding a lot of new rules and interfaces for pushdown nodes in the plan.
What about existing plugins? Does that mean we need rewrite logic for them to leverage this new mechanism?

**Comment by vdiravka** [comment]
> Calcite rules for adapters in most of the cases have parts of the logic related to the storage itself. My idea was to move out this logic from rules to "implementor", so we will have simple rules that could be reused, and adapters will provide only the "implementor" that will be responsible for checking whether the specific pushdown can be done and handling it.

Does that makes sense to continue this conversation in Calcite dev list?

**Comment by julianhyde** [comment]
I have no objections to moving the conversation to dev@calcite.

**Comment by vvysotskyi** [comment]
> We are adding a lot of new rules and interfaces for pushdown nodes in the plan.
What about existing plugins? Does that mean we need rewrite logic for them to leverage this new mechanism?

Existing plugins may be rewritten later if will be required to extend their functionality to support more pushdowns for them, but it makes sense to use it for newer plugins since it reduces the code amount that should be added to support pushdowns.

File: contrib/storage-mongo/src/main/java/org/apache/drill/exec/store/mongo/MongoAggregateUtils.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+      return s;
+    }
+    return quote(s);
+  }
+
+  public static String quote(String s) {
+    return "'" + s + "'";
+  }
+
+  private static boolean needsQuote(String s) {
```
[the remark this answers is not in the document]
Comment (vvysotskyi) [reply contrib/storage-mongo/src/main/java/org/apache/drill/exec/store/mongo/MongoAggregateUtils.java:65]: This class also contains other private methods.

**Comment by vvysotskyi** [comment]
> Does that makes sense to continue this conversation in Calcite dev list?

Yes, it makes sense to do that, but until Drill uses the older version of Calcite, this discussion wouldn't affect this project. So I propose to process this PR regardless of that discussion.

File: contrib/storage-mongo/src/main/java/org/apache/drill/exec/store/mongo/plan/RexToMongoTranslator.java (hunk 4 above)
Comment (vvysotskyi) [reply contrib/storage-mongo/src/main/java/org/apache/drill/exec/store/mongo/plan/RexToMongoTranslator.java:157]: Thanks, added

**Review by vdiravka — APPROVED** [review]
+1 LGTM

**Pull request merged**